packages/x-data-grid-pro/src/hooks/features/infiniteLoader/useGridInfiniteLoader.tsx

```diff
@@ -82,23 +87,35 @@ export const useGridInfiniteLoader = (
     }
   }, [virtualScroller, handleLoadMoreRows, isEnabled, marginBottom]);
 
+  const updateTarget = (node: HTMLElement | null) => {
+    if (triggerElement.current !== node) {
+      observer.current?.disconnect();
+
+      triggerElement.current = node;
+      if (triggerElement.current) {
+        observer.current?.observe(triggerElement.current);
+      }
+    }
+  };
+
   const triggerRef = React.useCallback(
     (node: HTMLElement | null) => {
       // Prevent the infite loading working in combination with lazy loading
       if (!isEnabled) {
         return;
       }
 
-      if (triggerElement.current !== node) {
-        observer.current?.disconnect();
-
-        triggerElement.current = node;
-        if (triggerElement.current) {
-          observer.current?.observe(triggerElement.current);
-        }
-      }
+      // If the user scrolls through the grid too fast it might happen that the observer is connected to the trigger element
+      // that will be intersecting the root inside the same render cycle (but not intersecting at the time of the connection).
+      // This will cause the observer to not call the callback with `isIntersecting` set to `true`.
+      // https://www.w3.org/TR/intersection-observer/#event-loop
+      // Delaying the connection to the next cycle helps since the observer will always call the callback the first time it is connected.
+      // https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/observe
+      // Related to
+      // https://github.com/mui/mui-x/issues/14116
+      updateTargetTimeout.start(0, () => updateTarget(node));
     },
-    [isEnabled],
+    [isEnabled, updateTargetTimeout],
```

Comment:
TIL `eslint-plugin-react-hooks` is smart enough not to require `updateTarget` as a dependency here because it doesn't rely on any state that might change 🤯